FCP-INDI/C-PAC
# 🐛 Differentiate sametime data and pipeline configs

## Fixes



Fixes #1776 by @shnizzedy
Related to #1774 & #1775 by @shnizzedy 

## Description

Before this PR, C-PAC would generate run-specific config files like
```BASH
cpac_data_config_{timestamp}.yml
cpac_pipeline_config_{timestamp}.yml
```

With this PR, those filenames include the first 8 characters of the hexadecimal digest of a hash of all `site_id`s, `subject_id`s, and `unique_id`s (like `site1,site2,site3_sub1,sub2,sub3_uid1,uid2,uid3`) in the given data config.

```BASH
cpac_data_config_{data_hash}_{timestamp}.yml
cpac_pipeline_config_{data_hash}_{timestamp}.yml
```

<a name="collision-note"></a>
**NOTE**: #1774 is still relevant if running multiple configs to the same top-level output directory (if the same data config launches multiple pipeline configs in the same second, this behavior will persist after this PR)

## Technical details

https://github.com/FCP-INDI/C-PAC/blob/68ac96ab2ccd9e509ad58ccabe5577b569b6b6bc/CPAC/utils/yaml_template.py#L320-L343 https://github.com/FCP-INDI/C-PAC/blob/68ac96ab2ccd9e509ad58ccabe5577b569b6b6bc/dev/docker_data/run.py#L710

## Tests

### Unit test
https://github.com/FCP-INDI/C-PAC/blob/68ac96ab2ccd9e509ad58ccabe5577b569b6b6bc/CPAC/utils/yaml_template.py#L334-L339
### Manual test
Start several runs in parallel (e.g., with SLURM) and see you have the same number of data configs and pipeline configs and runs, and that the `data_hash` values appear once each in a data config filename and once in a pipeline config filename (per pipeline config)

## Checklist


## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
